frontend/src/metabase/visualizations/echarts/cartesian/waterfall/model/dataset.ts

```diff
@@ -1,26 +1,72 @@
+import { t } from "ttag";
 import dayjs from "dayjs";
-import {
-  assertMultiMetricColumns,
-  type CartesianChartColumns,
-} from "metabase/visualizations/lib/graph/columns";
-import { isNumber } from "metabase/lib/types";
-
+import type { RowValue } from "metabase-types/api";
 import type { ComputedVisualizationSettings } from "metabase/visualizations/types";
-import type { XAxisModel } from "metabase/visualizations/echarts/cartesian/model/types";
+import type {
+  DataKey,
+  ChartDataset,
+  XAxisModel,
+  Datum,
+} from "metabase/visualizations/echarts/cartesian/model/types";
+import {
+  applySquareRootScaling,
+  replaceValues,
+} from "metabase/visualizations/echarts/cartesian/model/dataset";
 import { isAbsoluteDateTimeUnit } from "metabase-types/guards/date-time";
-import type { RowValues } from "metabase-types/api";
+
 import {
-  WATERFALL_EMPTY_VALUE,
-  type WaterfallDataset,
-  type WaterfallEmptyValue,
-} from "../types";
+  WATERFALL_DATA_KEYS,
+  WATERFALL_END_2_KEY,
+  WATERFALL_END_KEY,
+  WATERFALL_START_2_KEY,
+  WATERFALL_START_KEY,
+  WATERFALL_TOTAL_KEY,
+  WATERFALL_VALUE_KEY,
+} from "metabase/visualizations/echarts/cartesian/waterfall/constants";
+import { isNotNull, isNumber } from "metabase/lib/types";
+import { X_AXIS_DATA_KEY } from "metabase/visualizations/echarts/cartesian/constants/dataset";
+import { getNumberOr } from "metabase/visualizations/lib/settings/row-values";
+import { tryGetDate } from "../../utils/time-series";
+
+const replaceZerosForLogScale = (dataset: ChartDataset): ChartDataset => {
+  let hasZeros = false;
+  let minNonZeroValue = Infinity;
+
+  dataset.forEach(datum => {
+    const datumNumericValues = [
+      getNumberOr(datum[WATERFALL_START_KEY], null),
+      getNumberOr(datum[WATERFALL_END_KEY], null),
+    ].filter(isNotNull);
+
+    hasZeros = datumNumericValues.includes(0);
+
+    minNonZeroValue = Math.min(
+      minNonZeroValue,
+      ...datumNumericValues.filter(number => number !== 0),
+    );
+  });
+
+  if (!hasZeros && minNonZeroValue > 0) {
+    return dataset;
+  }
+
+  if (minNonZeroValue < 0) {
+    throw Error(t`X-axis must not cross 0 when using log scale.`);
```

Comment:
If we just throw an error the user will see the generic "Something's gone wrong" error which is not informative. Before we used to show a specific messaging saying that crossing 0 was incompatible with the waterfall chart.

<img width="1902" alt="Screenshot 2024-02-16 at 10 18 23 AM" src="https://github.com/metabase/metabase/assets/37751258/7843c136-40e5-46d2-a676-039a31c4cf8a">

<img width="1900" alt="Screenshot 2024-02-16 at 10 18 32 AM" src="https://github.com/metabase/metabase/assets/37751258/30d40ca0-d58e-4c8f-91e7-a8c88af5cf89">

Can we restore this error message? We can either do it in this PR or file an issue to do it later.

Reply:
I created an issue for this https://github.com/metabase/metabase/issues/38881